CA-40860: vhostmd config file needs to match changes to vhostmd
PR-1071: Formal Citrix support for vhostmd for SAP

--- ./vhostmd.xml	2010-04-07 18:40:06.000000000 +0100
+++ ./vhostmd.xml	2012-04-30 16:07:01.000000000 +0100
@@ -3,7 +3,8 @@
 
 <!--
 
-Copyright (C) 2008 Novell, Inc.
+Copyright (c) 2008 Novell, Inc.
+Copyright (c) 2009 Citrix Systems, Inc.
 
 Configuration file for virtual host metrics daemon (vhostmd).
 
@@ -14,82 +15,246 @@
 
 A metric's value is set to the output produced by executing its action.
 
-'action' can include the special token NAME, in which case the name of
-the vm currently under inspection is substituted for NAME.  Only useful
-within the vm element.
+For  metrics of vm context, the tokens NAME, VMID, and UUID may be used
+where these attributes of a VM are  normally  provided  in  a  command.
+When  the  metric is sampled, these tokens will be substituted with the
+actual name, ID, or UUID of the vm currently being sampled by  vhostmd.
+
 
 -->
 
-  <vhostmd>
-    <globals>
-      <disk>
-        <name>host-metrics-disk</name>
-        <path>/dev/shm/vhostmd0</path>
-        <size unit="k">256</size>
-      </disk>
-      <update_period>5</update_period>
-      <path>/usr/bin:/usr/sbin:/usr/share/vhostmd/scripts</path>
-      <transport>vbd</transport>
-      <!-- <transport>xenstore</transport> -->
-    </globals>
-    <metrics>
-      <metric type="string" context="host">
-        <name>HostName</name>
-        <action>virsh CONNECT hostname | tr -d '[:space:]'</action>
-      </metric>
-      <metric type="string" context="host">
-        <name>VirtualizationVendor</name>
-        <action>/bin/rpm -q --info xen | grep Vendor: | \
-                awk '{print substr($0, index($0,$5)) }'</action>
-      </metric>
-      <metric type="string" context="host">
-        <name>VirtualizationProductInfo</name>
-        <action>xm info | gawk '/^xen_(major|minor|extra)/ {print $3}' | \
-                tr -d . | tr '[:space:]' .</action>
-      </metric>
-      <metric type="uint32" context="host">
-        <name>TotalPhyCPUs</name>
-        <action>xm info | gawk '/^nr_cpus/ {print $3}'</action>
-      </metric>
-      <metric type="uint32" context="host">
-        <name>NumCPUs</name>
-        <action>xm info | gawk '/^nr_cpus/ {print $3}'</action>
-      </metric>
-      <metric type="uint64" context="host">
-        <name>TotalPhyMem</name>
-        <action>xm info | gawk '/^total_memory/ {print $3}'</action>
-      </metric>
-      <metric type="uint64" context="host">
-        <name>UsedMem</name>
-        <action>echo "$((`xentop -b -i 1 | gawk '/Domain-0/ {print $5}'` / 1024))"</action>
-      </metric>
-      <metric type="uint64" context="host">
-        <name>FreeMem</name>
-        <action>xm info | gawk '/^max_free_memory/ {print $3}'</action>
-      </metric>
-      <metric type="uint64" context="host">
-        <name>PagedInMemory</name>
-        <action>echo "$((`vmstat -s | gawk '/pages paged in/ {print $1}'` / 1024))"</action>
-      </metric>
-      <metric type="uint64" context="host">
-        <name>PagedOutMemory</name>
-        <action>echo "$((`vmstat -s | gawk '/pages paged out/ {print $1}'` / 1024))"</action>
-      </metric>
-      <metric type="group" context="host">
-        <name>PageRates</name>
-        <action>pagerate.py</action>
-        <variable name="PageInRate" type="uint64"/>
-        <variable name="PageFaultRate" type="uint64"/>
-      </metric>
-      <metric type="real64" context="host">
-        <name>TotalCPUTime</name>
-        <action>virsh CONNECT dominfo 0 | sed 's/: */:/' | \
-                gawk -F: '/CPU time/ {print $2;}'</action>
-      </metric>
-      <metric type="real64" context="vm">
-        <name>TotalCPUTime</name>
-        <action>virsh CONNECT dominfo NAME | sed 's/: */:/' | \
-                gawk -F: '/CPU time/ {print $2;}'</action>
-      </metric>
-    </metrics>
-  </vhostmd>
+<!--     1         2         3         4         5         6         -->
+<!-- 678901234567890123456789012345678901234567890123456789012345678 -->
+
+
+<vhostmd>
+  <globals>
+    <disk>
+      <name>host-metrics-disk</name>
+      <path>/dev/shm/vhostmd0</path>
+      <!-- must be between [128 KB, 256 MB] in size -->
+      <size unit="k">256</size>
+    </disk>
+    <update_period>60</update_period>
+    <!-- Do NOT use leading or trailing white space here  -->
+    <path>/bin:/sbin:/usr/bin:/usr/sbin:/usr/share/vhostmd/scripts:/opt/xensource/bin</path>
+    <transport>disk</transport>
+    <!-- <transport>vbd</transport>       -->
+    <!-- <transport>xenstore</transport>  -->
+  </globals>
+  <metrics>
+    <metric type="string" context="host">
+      <name>HostName</name>
+      <action>hostname</action>
+    </metric>
+    <metric type="uint64" context="host">
+      <name>Time</name>
+      <action>date +%s</action>
+    </metric>
+    <metric type="string" context="host">
+      <name>VirtualizationVendor</name>
+      <action>
+        printf "%s" "Citrix Systems, Inc."
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="string" context="host">
+      <name>VirtProductInfo</name>
+      <action>
+        xe host-param-get param-name=software-version             \
+                          param-key=product_version               \
+                          uuid=$(awk -F\' '/^INSTALLATION_UUID/   \
+                            {print $2}' /etc/xensource-inventory) \
+          | tr -d '\n'
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="string" context="host">
+      <name>HostSystemInfo</name>
+      <!-- hostname -s didn't work on xenserver -->
+      <action>hostname|sed 's/\..*$//'</action>
+    </metric>
+    <!-- SAP "XCIM" -->
+    <metric type="uint32" context="host">
+      <name>NumberOfPhysicalCPUs</name>
+      <!-- physical CPUs usable by the virtual machines -->
+      <!-- SAP "CIM" uses "NumberOfPhysicalCPUsUtilized"  -->
+      <!-- which means something different, i.e.          -->
+      <!--  "physical CPUs used by the virtual machines"  -->
+      <!-- but may be calculated (on client side) by      -->
+      <!-- Delta TotalCPUTime / Delta ElapsedTime         -->
+      <action>
+        xe host-cpu-list --minimal                                \
+          host-uuid=$(awk -F\' '/^INSTALLATION_UUID/ {print $2}'  \
+            /etc/xensource-inventory)                             \
+          | sed 's/,/ /g' | wc -w
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="uint64" context="host">
+      <name>MemoryAllocatedToVirtualServers</name>
+      <!-- physical memory used by the virtual machines,      -->
+      <!-- (_not_ physical memory usable by the virtual       -->
+      <!-- machines)                                          -->
+      <!--                                                    -->
+      <!-- USABLE: Calculation: TotalMem - Dom0Mem:
+      <action>
+        (
+          (xe host-param-get                                      \
+            uuid=$(awk -F\' '/^INSTALLATION_UUID/ {print $2}'     \
+            /etc/xensource-inventory) param-name=memory-total;    \
+            echo 1048576 / n) | dc;                               \
+          printf "%s" " ";                                        \
+          xentop -b -i 1 | gawk '/Domain-0/ {print $5 / 1024}';   \
+          printf " %s" "- n"
+        ) | dc
+      </action>
+      -->
+      <!--                                                    -->
+      <!-- USED: Calculation: TotalMem - FreeMem              -->
+      <action>
+        (
+          (xe host-param-get                                      \
+            uuid=$(awk -F\' '/^INSTALLATION_UUID/ {print $2}'     \
+            /etc/xensource-inventory) param-name=memory-total;    \
+            echo 1048576 / n) | dc;                               \
+          printf "%s" " ";                                        \
+          (xe host-param-get                                      \
+            uuid=$(awk -F\' '/^INSTALLATION_UUID/ {print $2}'     \
+            /etc/xensource-inventory) param-name=memory-free;     \
+            echo 1048576 / n) | dc;                               \
+          printf " %s" "- n"
+        ) | dc
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="uint64" context="host">
+      <name>FreePhysicalMemory</name>
+      <action>
+        (xe host-param-get                                        \
+          uuid=$(awk -F\' '/^INSTALLATION_UUID/ {print $2}'       \
+          /etc/xensource-inventory) param-name=memory-free;       \
+          echo 1048576 / n) | dc
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="uint64" context="host">
+      <name>FreeVirtualMemory</name>
+      <!-- Calculation: Equal to FreePhysicalMemory           -->
+      <action>
+        (xe host-param-get                                        \
+          uuid=$(awk -F\' '/^INSTALLATION_UUID/ {print $2}'       \
+          /etc/xensource-inventory) param-name=memory-free;       \
+          echo 1048576 / n) | dc
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="uint64" context="host">
+      <name>UsedVirtualMemory</name>
+      <!-- Calculation: TotalMem - FreeMem                    -->
+      <action>
+        (
+          (xe host-param-get                                      \
+            uuid=$(awk -F\' '/^INSTALLATION_UUID/ {print $2}'     \
+            /etc/xensource-inventory) param-name=memory-total;    \
+            echo 1048576 / n) | dc;                               \
+          printf "%s" " ";                                        \
+          (xe host-param-get                                      \
+            uuid=$(awk -F\' '/^INSTALLATION_UUID/ {print $2}'     \
+            /etc/xensource-inventory) param-name=memory-free;     \
+            echo 1048576 / n) | dc;                               \
+          printf " %s" "- n"
+        ) | dc
+      </action>
+    </metric>
+    <!-- SAP "XCIM" -->
+    <metric type="uint64" context="host">
+      <name>PagedInMemory</name>
+      <action>
+        vmstat -s | awk 'BEGIN {
+                        cmd = "getconf PAGESIZE";
+                        cmd | getline pagesize;
+                        close(cmd);
+                      }
+                      /pages swapped in/ {
+                        printf "%d\n", $1 / 1024 * pagesize / 1024;
+                      }'
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="uint64" context="host">
+      <name>PagedOutMemory</name>
+      <action>
+        vmstat -s | awk 'BEGIN {
+                        cmd = "getconf PAGESIZE";
+                        cmd | getline pagesize;
+                        close(cmd);
+                      }
+                      /pages swapped out/ {
+                        printf "%d\n", $1 / 1024 * pagesize / 1024;
+                      }'
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="real64" context="host">
+      <name>TotalCPUTime</name>
+      <action>
+        list_domains -all                           \
+          |tail -n +2                               \
+          |sed -e 's/^[ \t]*//' -e 's/[ \t]*$//'    \
+          |awk -F'[ \t]*[|][ \t]*' '{ print $9 }'   \
+          |tr -c -d '[:digit:]\n'                   \
+          |awk '{ n = $1/1000000000.0 + n } END { printf "%f", n; }'
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="real64" context="vm">
+      <name>TotalCPUTime</name>
+      <action>
+        list_domains -all -domid VMID               \
+          |tail -n +2                               \
+          |sed -e 's/^[ \t]*//' -e 's/[ \t]*$//'    \
+          |awk -F'[ \t]*[|][ \t]*' '{ print $9 }'   \
+          |tr -c -d '[:digit:]'                     \
+          |awk '{ printf "%f", $1/1000000000.0; }'
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="uint32" context="vm">
+      <name>ResourceProcessorLimit</name>
+      <action>
+        list_domains -all -domid VMID               \
+          |tail -n +2                               \
+          |sed -e 's/^[ \t]*//' -e 's/[ \t]*$//'    \
+          |awk -F'[ \t]*[|][ \t]*' '{ print $10 }'  \
+          |tr -c -d '[:digit:]'
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="uint64" context="vm">
+      <name>ResourceMemoryLimit</name>
+      <action>
+        list_domains -all -domid VMID               \
+          |tail -n +2                               \
+          |sed -e 's/^[ \t]*//' -e 's/[ \t]*$//'    \
+          |awk -F'[ \t]*[|][ \t]*' '{ print $5 }'   \
+          |tr -c -d '[:digit:]'
+      </action>
+    </metric>
+    <!-- SAP "CIM"  -->
+    <metric type="uint64" context="vm">
+      <name>PhysicalMemoryAllocatedToVirtualSystem</name>
+      <action>
+        list_domains -all -domid VMID               \
+          |tail -n +2                               \
+          |sed -e 's/^[ \t]*//' -e 's/[ \t]*$//'    \
+          |awk -F'[ \t]*[|][ \t]*' '{ print $4 }'   \
+          |tr -c -d '[:digit:]'
+      </action>
+    </metric>
+  </metrics>
+</vhostmd>
+<!--
+  vi:ts=2:sw=2:expandtab:ignorecase:nu:ruler
+-->
+<!-- EOF -->
